In [1]:
import numpy as np # library to handle data in a vectorized manner
import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
import json # library to handle JSON files
import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes 
# uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

#!conda install -c conda-forge geopy --yes 
!from geopy.geocoders import Nominatim 
# module to convert an address into latitude and longitude values
print('Libraries imported.')

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - folium=0.5.0


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    ca-certificates-2019.6.16  |       hecc5488_0         145 KB  conda-forge
    certifi-2019.6.16          |           py36_1         149 KB  conda-forge
    vincent-0.4.4              |             py_1          28 KB  conda-forge
    branca-0.3.1               |             py_0          25 KB  conda-forge
    folium-0.5.0               |             py_0          45 KB  conda-forge
    altair-3.2.0               |           py36_0         770 KB  conda-forge
    openssl-1.1.1c             |       h516909a_0         2.1 MB  conda-forge
    ------------------------------------------------------------
                                           Total:         3.3 MB

The following NEW packages will be 

### Scraping of Table
Import the HTML using the requests library, and then pass it through as a BeautifulSoup object.
Make use of soup.find() to search for  tables on the page, and choose the item from the list to read as a dataframe.

### Cleaning up Table
Drop the Borough's labeled as "Not assigned" and reset the index to fix the row numbering. And reassign the values of the "Not assigned" neighborhoods to be the name of the Borough.

In [3]:
import urllib.request
import bs4 as bs

In [4]:
url=urllib.request.urlopen('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M').read()
soup=bs.BeautifulSoup(url, features="lxml")
# print(soup)
test=soup.find('table', {"class":"wikitable"})
# print(test)
df=pd.read_html(str(test))
df=df[0]
df=df.loc[df['Borough'] != "Not assigned"]
df = df.reset_index(drop=True)
df.Neighbourhood[df.Neighbourhood=='Not assigned'] = df.Borough
df

Postcode           Borough  \
0        M3A        North York   
1        M4A        North York   
2        M5A  Downtown Toronto   
3        M5A  Downtown Toronto   
4        M6A        North York   
5        M6A        North York   
6        M7A      Queen's Park   
7        M9A         Etobicoke   
8        M1B       Scarborough   
9        M1B       Scarborough   
10       M3B        North York   
11       M4B         East York   
12       M4B         East York   
13       M5B  Downtown Toronto   
14       M5B  Downtown Toronto   
15       M6B        North York   
16       M9B         Etobicoke   
17       M9B         Etobicoke   
18       M9B         Etobicoke   
19       M9B         Etobicoke   
20       M9B         Etobicoke   
21       M1C       Scarborough   
22       M1C       Scarborough   
23       M1C       Scarborough   
24       M3C        North York   
25       M3C        North York   
26       M4C         East York   
27       M5C  Downtown Toronto   
28       M6C              York   
29       M9C         Etobicoke   
30       M9C         Etobicoke   
31       M9C         Etobicoke   
32       M9C         Etobicoke   
33       M1E       Scarborough   
34       M1E       Scarborough   
35       M1E       Scarborough   
36       M4E      East Toronto   
37       M5E  Downtown Toronto   
38       M6E              York   
39       M1G       Scarborough   
40       M4G         East York   
41       M5G  Downtown Toronto   
42       M6G  Downtown Toronto   
43       M1H       Scarborough   
44       M2H        North York   
45       M3H        North York   
46       M3H        North York   
47       M3H        North York   
48       M4H         East York   
49       M5H  Downtown Toronto   
50       M5H  Downtown Toronto   
51       M5H  Downtown Toronto   
52       M6H      West Toronto   
53       M6H      West Toronto   
54       M1J       Scarborough   
55       M2J        North York   
56       M2J        North York   
57       M2J        North York   
58       M3J        North York   
59       M3J        North York   
60       M4J         East York   
61       M5J  Downtown Toronto   
62       M5J  Downtown Toronto   
63       M5J  Downtown Toronto   
64       M6J      West Toronto   
65       M6J      West Toronto   
66       M1K       Scarborough   
67       M1K       Scarborough   
68       M1K       Scarborough   
69       M2K        North York   
70       M3K        North York   
71       M3K        North York   
72       M4K      East Toronto   
73       M4K      East Toronto   
74       M5K  Downtown Toronto   
75       M5K  Downtown Toronto   
76       M6K      West Toronto   
77       M6K      West Toronto   
78       M6K      West Toronto   
79       M1L       Scarborough   
80       M1L       Scarborough   
81       M1L       Scarborough   
82       M2L        North York   
83       M2L        North York   
84       M3L        North York   
85       M4L      East Toronto   
86       M4L      East Toronto   
87       M5L  Downtown Toronto   
88       M5L  Downtown Toronto   
89       M6L        North York   
90       M6L        North York   
91       M6L        North York   
92       M9L        North York   
93       M1M       Scarborough   
94       M1M       Scarborough   
95       M1M       Scarborough   
96       M2M        North York   
97       M2M        North York   
98       M3M        North York   
99       M4M      East Toronto   
100      M5M        North York   
101      M5M        North York   
102      M6M              York   
103      M6M              York   
104      M6M              York   
105      M6M              York   
106      M9M        North York   
107      M9M        North York   
108      M1N       Scarborough   
109      M1N       Scarborough   
110      M2N        North York   
111      M3N        North York   
112      M4N   Central Toronto   
113      M5N   Central Toronto   
114      M6N              York   
115      M6N              York   
116      M9N              

### Combine Neighborhoods
Use groupby method and combine neighborhood cells based on matching postcode/boroughs, and again, reset the index to fix the row numbering.

In [5]:
df = df.groupby(['Postcode','Borough'])['Neighbourhood'].apply(', '.join).reset_index()
df

Postcode           Borough  \
0        M1B       Scarborough   
1        M1C       Scarborough   
2        M1E       Scarborough   
3        M1G       Scarborough   
4        M1H       Scarborough   
5        M1J       Scarborough   
6        M1K       Scarborough   
7        M1L       Scarborough   
8        M1M       Scarborough   
9        M1N       Scarborough   
10       M1P       Scarborough   
11       M1R       Scarborough   
12       M1S       Scarborough   
13       M1T       Scarborough   
14       M1V       Scarborough   
15       M1W       Scarborough   
16       M1X       Scarborough   
17       M2H        North York   
18       M2J        North York   
19       M2K        North York   
20       M2L        North York   
21       M2M        North York   
22       M2N        North York   
23       M2P        North York   
24       M2R        North York   
25       M3A        North York   
26       M3B        North York   
27       M3C        North York   
28       M3H        North York   
29       M3J        North York   
30       M3K        North York   
31       M3L        North York   
32       M3M        North York   
33       M3N        North York   
34       M4A        North York   
35       M4B         East York   
36       M4C         East York   
37       M4E      East Toronto   
38       M4G         East York   
39       M4H         East York   
40       M4J         East York   
41       M4K      East Toronto   
42       M4L      East Toronto   
43       M4M      East Toronto   
44       M4N   Central Toronto   
45       M4P   Central Toronto   
46       M4R   Central Toronto   
47       M4S   Central Toronto   
48       M4T   Central Toronto   
49       M4V   Central Toronto   
50       M4W  Downtown Toronto   
51       M4X  Downtown Toronto   
52       M4Y  Downtown Toronto   
53       M5A  Downtown Toronto   
54       M5B  Downtown Toronto   
55       M5C  Downtown Toronto   
56       M5E  Downtown Toronto   
57       M5G  Downtown Toronto   
58       M5H  Downtown Toronto   
59       M5J  Downtown Toronto   
60       M5K  Downtown Toronto   
61       M5L  Downtown Toronto   
62       M5M        North York   
63       M5N   Central Toronto   
64       M5P   Central Toronto   
65       M5R   Central Toronto   
66       M5S  Downtown Toronto   
67       M5T  Downtown Toronto   
68       M5V  Downtown Toronto   
69       M5W  Downtown Toronto   
70       M5X  Downtown Toronto   
71       M6A        North York   
72       M6B        North York   
73       M6C              York   
74       M6E              York   
75       M6G  Downtown Toronto   
76       M6H      West Toronto   
77       M6J      West Toronto   
78       M6K      West Toronto   
79       M6L        North York   
80       M6M              York   
81       M6N              York   
82       M6P      West Toronto   
83       M6R      West Toronto   
84       M6S      West Toronto   
85       M7A      Queen's Park   
86       M7R       Mississauga   
87       M7Y      East Toronto   
88       M8V         Etobicoke   
89       M8W         Etobicoke   
90       M8X         Etobicoke   
91       M8Y         Etobicoke   
92       M8Z         Etobicoke   
93       M9A         Etobicoke   
94       M9B         Etobicoke   
95       M9C         Etobicoke   
96       M9L        North York   
97       M9M        North York   
98       M9N              York   
99       M9P         Etobicoke   
100      M9R         Etobicoke   
101      M9V         Etobicoke   
102      M9W         Etobicoke   

                                         Neighbourhood  
0                                       Rouge, Malvern  
1               Highland Creek, Rouge Hill, Port Union  
2                    Guildwood, Morningside, West Hill  
3                                               Woburn  
4                                            Cedarbrae  
5                                  Scarborough Village  
6          East Birchmount Park, Ionview, Kennedy Park  
7          

### Shape Method
Using the df.shape method, we see there are 103 rows and 3 columns.

In [6]:
df.shape

(103, 3)

### Imporitng the CSV
Import the provided CSV file, naming it "geolocations".

In [8]:
geolocations = pd.read_csv("https://cocl.us/Geospatial_data")

### Adding the Latitude and Longitude
Perform a join with the geolocations dataframe's index set to be the postal code, on the original dataframe's Postcode column. 

In [9]:
df = df.join(geolocations.set_index('Postal Code'), on='Postcode')
df

Postcode           Borough  \
0        M1B       Scarborough   
1        M1C       Scarborough   
2        M1E       Scarborough   
3        M1G       Scarborough   
4        M1H       Scarborough   
5        M1J       Scarborough   
6        M1K       Scarborough   
7        M1L       Scarborough   
8        M1M       Scarborough   
9        M1N       Scarborough   
10       M1P       Scarborough   
11       M1R       Scarborough   
12       M1S       Scarborough   
13       M1T       Scarborough   
14       M1V       Scarborough   
15       M1W       Scarborough   
16       M1X       Scarborough   
17       M2H        North York   
18       M2J        North York   
19       M2K        North York   
20       M2L        North York   
21       M2M        North York   
22       M2N        North York   
23       M2P        North York   
24       M2R        North York   
25       M3A        North York   
26       M3B        North York   
27       M3C        North York   
28       M3H        North York   
29       M3J        North York   
30       M3K        North York   
31       M3L        North York   
32       M3M        North York   
33       M3N        North York   
34       M4A        North York   
35       M4B         East York   
36       M4C         East York   
37       M4E      East Toronto   
38       M4G         East York   
39       M4H         East York   
40       M4J         East York   
41       M4K      East Toronto   
42       M4L      East Toronto   
43       M4M      East Toronto   
44       M4N   Central Toronto   
45       M4P   Central Toronto   
46       M4R   Central Toronto   
47       M4S   Central Toronto   
48       M4T   Central Toronto   
49       M4V   Central Toronto   
50       M4W  Downtown Toronto   
51       M4X  Downtown Toronto   
52       M4Y  Downtown Toronto   
53       M5A  Downtown Toronto   
54       M5B  Downtown Toronto   
55       M5C  Downtown Toronto   
56       M5E  Downtown Toronto   
57       M5G  Downtown Toronto   
58       M5H  Downtown Toronto   
59       M5J  Downtown Toronto   
60       M5K  Downtown Toronto   
61       M5L  Downtown Toronto   
62       M5M        North York   
63       M5N   Central Toronto   
64       M5P   Central Toronto   
65       M5R   Central Toronto   
66       M5S  Downtown Toronto   
67       M5T  Downtown Toronto   
68       M5V  Downtown Toronto   
69       M5W  Downtown Toronto   
70       M5X  Downtown Toronto   
71       M6A        North York   
72       M6B        North York   
73       M6C              York   
74       M6E              York   
75       M6G  Downtown Toronto   
76       M6H      West Toronto   
77       M6J      West Toronto   
78       M6K      West Toronto   
79       M6L        North York   
80       M6M              York   
81       M6N              York   
82       M6P      West Toronto   
83       M6R      West Toronto   
84       M6S      West Toronto   
85       M7A      Queen's Park   
86       M7R       Mississauga   
87       M7Y      East Toronto   
88       M8V         Etobicoke   
89       M8W         Etobicoke   
90       M8X         Etobicoke   
91       M8Y         Etobicoke   
92       M8Z         Etobicoke   
93       M9A         Etobicoke   
94       M9B         Etobicoke   
95       M9C         Etobicoke   
96       M9L        North York   
97       M9M        North York   
98       M9N              York   
99       M9P         Etobicoke   
100      M9R         Etobicoke   
101      M9V         Etobicoke   
102      M9W         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                       Rouge, Malvern  43.806686 -79.194353  
1               Highland Creek, Rouge Hill, Port Union  43.784535 -79.160497  
2                    Guildwood, Morningside, West Hill  43.763573 -79.188711  
3                                               Woburn  43.770992 -79.216917  
4                                            Cedarbrae  43.773136 -79.23

In [10]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(df['Borough'].unique()),
        df.shape[0]
    )
)

The dataframe has 11 boroughs and 103 neighborhoods.


## Analyse Neighbourhoods

In [11]:
#!conda install -c conda-forge geopy --yes 
from geopy.geocoders import Nominatim # module to convert an address into latitude and longitude values

Solving environment: done

# All requested packages already installed.



In [12]:
address = 'Toronto, Canada'
geolocator = Nominatim(user_agent="ca_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.653963, -79.387207.


### Mapping the neighborhoods.

In [14]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(df['Latitude'], df['Longitude'], df['Borough'], df['Neighbourhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

In [15]:
CLIENT_ID = '4WOAAQII3QSIJH1NQLJAKLR3CVU2CXZQAJI215NWI3223KXN' # your Foursquare ID
CLIENT_SECRET = 'BJQS0NNC5Y3IG2NTRFTCKFGMDYSQPEILPHIG03Y0SMTZL0FW' # your Foursquare Secret
VERSION = '20180605'
LIMIT = 30
print('Deepak Khare :')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Deepak Khare :
CLIENT_ID: 4WOAAQII3QSIJH1NQLJAKLR3CVU2CXZQAJI215NWI3223KXN
CLIENT_SECRET:BJQS0NNC5Y3IG2NTRFTCKFGMDYSQPEILPHIG03Y0SMTZL0FW


## Focus in on One Borough
Narrow down the focus to Downtown Toronto, like how we narrowed down to just Manhattan from NYC in the lab.

In [17]:
downtown_data = df[df['Borough'] == 'Downtown Toronto'].reset_index(drop=True)
downtown_data.head()

Postcode           Borough                Neighbourhood   Latitude  \
0      M4W  Downtown Toronto                     Rosedale  43.679563   
1      M4X  Downtown Toronto  Cabbagetown, St. James Town  43.667967   
2      M4Y  Downtown Toronto         Church and Wellesley  43.665860   
3      M5A  Downtown Toronto    Harbourfront, Regent Park  43.654260   
4      M5B  Downtown Toronto     Ryerson, Garden District  43.657162   

   Longitude  
0 -79.377529  
1 -79.367675  
2 -79.383160  
3 -79.360636  
4 -79.378937

In [18]:
address = 'Downtown Toronto, ON'
geolocator = Nominatim(user_agent="ca_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Downtown Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Downtown Toronto are 43.655115, -79.380219.


### Map downtown Toronto

In [19]:
# create map of Manhattan using latitude and longitude values
map_downtown = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(downtown_data['Latitude'], downtown_data['Longitude'], downtown_data['Neighbourhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_downtown)  
    
map_downtown

In [23]:
df.loc[0,'Neighbourhood']

'Rouge, Malvern'

In [24]:
neighbourhood_latitude = downtown_data.loc[0, 'Latitude'] # neighborhood latitude value
neighbourhood_longitude = downtown_data.loc[0, 'Longitude'] # neighborhood longitude value

neighbourhood_name = downtown_data.loc[0, 'Neighbourhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighbourhood_name, 
                                                               neighbourhood_latitude, 
                                                               neighbourhood_longitude))

Latitude and longitude values of Rosedale are 43.6795626, -79.37752940000001.


### Generate API URL, and perform GET request

In [26]:
LIMIT = 100 # limit of number of venues returned by Foursquare API
radius = 500 # define radius
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighbourhood_latitude, 
    neighbourhood_longitude, 
    radius, 
    LIMIT)
url # display URL

'https://api.foursquare.com/v2/venues/explore?&client_id=4WOAAQII3QSIJH1NQLJAKLR3CVU2CXZQAJI215NWI3223KXN&client_secret=BJQS0NNC5Y3IG2NTRFTCKFGMDYSQPEILPHIG03Y0SMTZL0FW&v=20180605&ll=43.6795626,-79.37752940000001&radius=500&limit=100'

In [27]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5d625f8a018cbb002cdb6cff'},
 'response': {'headerLocation': 'Rosedale',
  'headerFullLocation': 'Rosedale, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 5,
  'suggestedBounds': {'ne': {'lat': 43.6840626045, 'lng': -79.37131878274371},
   'sw': {'lat': 43.675062595499995, 'lng': -79.38374001725632}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4bae2150f964a520df873be3',
       'name': 'Mooredale House',
       'location': {'address': '146 Crescent Rd.',
        'crossStreet': 'btwn. Lamport Ave. and Mt. Pleasant Rd.',
        'lat': 43.678630645646535,
        'lng': -79.38009142511322,
        'labeledLatLngs': [{'label': 'display',
          'lat': 43.678630645646535,
          'lng': -79.380091425113

### Define function to extracts category of the venue

In [28]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [29]:
venues = results['response']['groups'][0]['items']    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]
nearby_venues.head()

name  categories        lat        lng
0       Mooredale House    Building  43.678631 -79.380091
1         Rosedale Park  Playground  43.682328 -79.378934
2          Whitney Park        Park  43.682036 -79.373788
3  Alex Murray Parkette        Park  43.678300 -79.382773
4        Milkman's Lane       Trail  43.676352 -79.373842

In [30]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

5 venues were returned by Foursquare.


### Retrieve all neighborhoods in Toronto using same process

In [31]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [32]:
downtown_venues = getNearbyVenues(names=downtown_data['Neighbourhood'],
                                   latitudes=downtown_data['Latitude'],
                                   longitudes=downtown_data['Longitude']
                                  )

Rosedale
Cabbagetown, St. James Town
Church and Wellesley
Harbourfront, Regent Park
Ryerson, Garden District
St. James Town
Berczy Park
Central Bay Street
Adelaide, King, Richmond
Harbourfront East, Toronto Islands, Union Station
Design Exchange, Toronto Dominion Centre
Commerce Court, Victoria Hotel
Harbord, University of Toronto
Chinatown, Grange Park, Kensington Market
CN Tower, Bathurst Quay, Island airport, Harbourfront West, King and Spadina, Railway Lands, South Niagara
Stn A PO Boxes 25 The Esplanade
First Canadian Place, Underground city
Christie


In [33]:
print(downtown_venues.shape)
downtown_venues.head()

(1284, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0     Rosedale              43.679563              -79.377529   
1     Rosedale              43.679563              -79.377529   
2     Rosedale              43.679563              -79.377529   
3     Rosedale              43.679563              -79.377529   
4     Rosedale              43.679563              -79.377529   

                  Venue  Venue Latitude  Venue Longitude Venue Category  
0       Mooredale House       43.678631       -79.380091       Building  
1         Rosedale Park       43.682328       -79.378934     Playground  
2          Whitney Park       43.682036       -79.373788           Park  
3  Alex Murray Parkette       43.678300       -79.382773           Park  
4        Milkman's Lane       43.676352       -79.373842          Trail

In [34]:
downtown_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Adelaide, King, Richmond                                              100   
Berczy Park                                                            57   
CN Tower, Bathurst Quay, Island airport, Harbou...                     14   
Cabbagetown, St. James Town                                            45   
Central Bay Street                                                     84   
Chinatown, Grange Park, Kensington Market                             100   
Christie                                                               15   
Church and Wellesley                                                   84   
Commerce Court, Victoria Hotel                                        100   
Design Exchange, Toronto Dominion Centre                              100   
First Canadian Place, Underground city                                100   
Harbord, University of Toronto                                         35   
Harbourfront East, Toronto Islands, Union Station                     100   
Harbourfront, Regent Park                                              49   
Rosedale                                                                5   
Ryerson, Garden District                                              100   
St. James Town                                                        100   
Stn A PO Boxes 25 The Esplanade                                        96   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Adelaide, King, Richmond                                               100   
Berczy Park                                                             57   
CN Tower, Bathurst Quay, Island airport, Harbou...                      14   
Cabbagetown, St. James Town                                             45   
Central Bay Street                                                      84   
Chinatown, Grange Park, Kensington Market                              100   
Christie                                                                15   
Church and Wellesley                                                    84   
Commerce Court, Victoria Hotel                                         100   
Design Exchange, Toronto Dominion Centre                               100   
First Canadian Place, Underground city                                 100   
Harbord, University of Toronto                                          35   
Harbourfront East, Toronto Islands, Union Station                      100   
Harbourfront, Regent Park                                               49   
Rosedale                                                                 5   
Ryerson, Garden District                                               100   
St. James Town                                                         100   
Stn A PO Boxes 25 The Esplanade                                         96   

                                                    Venue  Venue Latitude  \
Neighborhood                                                                
Adelaide, King, Richmond                              100             100   
Berczy Park                                            57              57   
CN Tower, Bathurst Quay, Island airport, Harbou...     14              14   
Cabbagetown, St. James Town                            45              45   
Central Bay Street                                     84              84   
Chinatown, Grange Park, Kensington Market             100             100   
Christie                                               15              15   
Church and Wellesley                                   84              84   
Commerce Court, Victoria Hotel                        100             100   
Design Exchange, Toronto Dominion Centre              100             100   
First Canadian Place, Unde

In [35]:
print('There are {} uniques categories.'.format(len(downtown_venues['Venue Category'].unique())))

There are 207 uniques categories.


### Analyze each Neighborhood

In [36]:
# one hot encoding
downtown_onehot = pd.get_dummies(downtown_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
downtown_onehot['Neighborhood'] = downtown_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [downtown_onehot.columns[-1]] + list(downtown_onehot.columns[:-1])
downtown_onehot = downtown_onehot[fixed_columns]

downtown_onehot.head()

Yoga Studio  Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0            0                  0        0                   0             0   
1            0                  0        0                   0             0   
2            0                  0        0                   0             0   
3            0                  0        0                   0             0   
4            0                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  \
0             0         0            0           0                    0   
1             0         0            0           0                    0   
2             0         0            0           0                    0   
3             0         0            0           0                    0   
4             0         0            0           0                    0   

   Asian Restaurant  BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  \
0                 0          0           0           0       0     0    0   
1                 0          0           0           0       0     0    0   
2                 0          0           0           0       0     0    0   
3                 0          0           0           0       0     0    0   
4                 0          0           0           0       0     0    0   

   Baseball Stadium  Basketball Stadium  Beach  Beer Bar  Beer Store  Bistro  \
0                 0                   0      0         0           0       0   
1                 0                   0      0         0           0       0   
2                 0                   0      0         0           0       0   
3                 0                   0      0         0           0       0   
4                 0                   0      0         0           0       0   

   Boat or Ferry  Bookstore  Brazilian Restaurant  Breakfast Spot  Brewery  \
0              0          0                     0               0        0   
1              0          0                     0               0        0   
2              0          0                     0               0        0   
3              0          0                     0               0        0   
4              0          0                     0               0        0   

   Bubble Tea Shop  Building  Burger Joint  Burrito Place  Butcher  Café  \
0                0         1             0              0        0     0   
1                0         0             0              0        0     0   
2                0         0             0              0        0     0   
3                0         0             0              0        0     0   
4                0         0             0              0        0     0   

   Camera Store  Caribbean Restaurant  Cheese Shop  Chinese Restaurant  \
0             0                     0            0                   0   
1             0                     0            0                   0   
2             0                     0            0                   0   
3             0                     0            0                   0   
4             0                     0            0                   0   

   Chocolate Shop  Church  Clothing Store  Cocktail Bar  Coffee Shop  \
0               0       0               0             0            0   
1               0       0               0             0            0   
2               0       0               0             0            0   
3    

### Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category and then let us confirm the new size

In [37]:
dt_grouped = downtown_onehot.groupby('Neighborhood').mean().reset_index()
dt_grouped

Neighborhood  Yoga Studio  \
0                            Adelaide, King, Richmond     0.000000   
1                                         Berczy Park     0.000000   
2   CN Tower, Bathurst Quay, Island airport, Harbo...     0.000000   
3                         Cabbagetown, St. James Town     0.000000   
4                                  Central Bay Street     0.011905   
5           Chinatown, Grange Park, Kensington Market     0.000000   
6                                            Christie     0.000000   
7                                Church and Wellesley     0.011905   
8                      Commerce Court, Victoria Hotel     0.000000   
9            Design Exchange, Toronto Dominion Centre     0.000000   
10             First Canadian Place, Underground city     0.000000   
11                     Harbord, University of Toronto     0.028571   
12  Harbourfront East, Toronto Islands, Union Station     0.000000   
13                          Harbourfront, Regent Park     0.000000   
14                                           Rosedale     0.000000   
15                           Ryerson, Garden District     0.000000   
16                                     St. James Town     0.000000   
17                    Stn A PO Boxes 25 The Esplanade     0.000000   

    Afghan Restaurant   Airport  Airport Food Court  Airport Gate  \
0            0.000000  0.000000            0.000000      0.000000   
1            0.000000  0.000000            0.000000      0.000000   
2            0.000000  0.071429            0.071429      0.071429   
3            0.000000  0.000000            0.000000      0.000000   
4            0.000000  0.000000            0.000000      0.000000   
5            0.000000  0.000000            0.000000      0.000000   
6            0.000000  0.000000            0.000000      0.000000   
7            0.011905  0.000000            0.000000      0.000000   
8            0.000000  0.000000            0.000000      0.000000   
9            0.000000  0.000000            0.000000      0.000000   
10           0.000000  0.000000            0.000000      0.000000   
11           0.000000  0.000000            0.000000      0.000000   
12           0.000000  0.000000            0.000000      0.000000   
13           0.000000  0.000000            0.000000      0.000000   
14           0.000000  0.000000            0.000000      0.000000   
15           0.000000  0.000000            0.000000      0.000000   
16           0.000000  0.000000            0.000000      0.000000   
17           0.000000  0.000000            0.000000      0.000000   

    Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0         0.000000         0.000000          0.000000             0.030000   
1         0.000000         0.000000          0.000000             0.000000   
2         0.142857         0.142857          0.142857             0.000000   
3         0.000000         0.000000          0.000000             0.000000   
4         0.000000         0.000000          0.000000             0.011905   
5         0.000000         0.000000          0.000000             0.000000   
6         0.000000         0.000000          0.000000             0.000000   
7         0.000000         0.000000          0.000000             0.011905   
8         0.000000         0.000000          0.000000             0.040000   
9         0.000000         0.000000          0.000000             0.030000   
10        0.000000         0.000000          0.000000             0.030000   
11        0.000000         0.000000          0.000000             0.000000   
12        0.000000         0.000000          0.000000             0.000000   
13        0.000000         0.000000          0.000000             0.000000   
14        0.000000         0.000000          0.000000             0.000000   
15        0.000000         0.000000          0.000000             0.010000   
16        0.000000         0.000000          0.000000             0.0200

In [38]:
dt_grouped.shape

(18, 207)

### Let's print each neighborhood along with the top 5 most common venues

In [39]:
num_top_venues = 5

for hood in dt_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = dt_grouped[dt_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Adelaide, King, Richmond----
             venue  freq
0      Coffee Shop  0.08
1             Café  0.05
2       Steakhouse  0.04
3              Bar  0.04
4  Thai Restaurant  0.04


----Berczy Park----
            venue  freq
0     Coffee Shop  0.09
1    Cocktail Bar  0.05
2  Farmers Market  0.04
3            Café  0.04
4        Beer Bar  0.04


----CN Tower, Bathurst Quay, Island airport, Harbourfront West, King and Spadina, Railway Lands, South Niagara----
                venue  freq
0      Airport Lounge  0.14
1     Airport Service  0.14
2    Airport Terminal  0.14
3             Airport  0.07
4  Airport Food Court  0.07


----Cabbagetown, St. James Town----
         venue  freq
0         Park  0.07
1  Coffee Shop  0.07
2          Pub  0.04
3         Café  0.04
4   Restaurant  0.04


----Central Bay Street----
                venue  freq
0         Coffee Shop  0.14
1  Italian Restaurant  0.05
2      Ice Cream Shop  0.05
3                Café  0.04
4      Sandwich Place  0.04


---

In [40]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [41]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = dt_grouped['Neighborhood']

for ind in np.arange(dt_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(dt_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                           Adelaide, King, Richmond           Coffee Shop   
1                                        Berczy Park           Coffee Shop   
2  CN Tower, Bathurst Quay, Island airport, Harbo...        Airport Lounge   
3                        Cabbagetown, St. James Town           Coffee Shop   
4                                 Central Bay Street           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Café                   Bar            Steakhouse   
1          Cocktail Bar                  Café            Steakhouse   
2      Airport Terminal       Airport Service         Boat or Ferry   
3                  Park                Bakery                  Café   
4    Italian Restaurant        Ice Cream Shop        Sandwich Place   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0       Thai Restaurant                   Gym                 Hotel   
1              Beer Bar    Seafood Restaurant                Bakery   
2      Sculpture Garden                 Plane           Coffee Shop   
3                   Pub     Convenience Store    Italian Restaurant   
4          Burger Joint                  Café            Restaurant   

       8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0        American Restaurant            Restaurant         Breakfast Spot  
1                Cheese Shop        Farmers Market              Irish Pub  
2               Airport Gate       Harbor / Marina                Airport  
3                Pizza Place            Restaurant          Deli / Bodega  
4  Middle Eastern Restaurant                   Spa                 Bakery

##  Cluster Neighborhoods
### Run k-means to cluster the neighborhood into 5 clusters.

In [42]:
# set number of clusters
kclusters = 5

dt_grouped_clustering = dt_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(dt_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 3, 0, 0, 2, 4, 0, 0, 0], dtype=int32)

### Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [43]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

dt_merged = downtown_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
dt_merged = dt_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighbourhood')

dt_merged.head() # check the last columns!

Postcode           Borough                Neighbourhood   Latitude  \
0      M4W  Downtown Toronto                     Rosedale  43.679563   
1      M4X  Downtown Toronto  Cabbagetown, St. James Town  43.667967   
2      M4Y  Downtown Toronto         Church and Wellesley  43.665860   
3      M5A  Downtown Toronto    Harbourfront, Regent Park  43.654260   
4      M5B  Downtown Toronto     Ryerson, Garden District  43.657162   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.377529               1                  Park            Playground   
1 -79.367675               0           Coffee Shop                  Park   
2 -79.383160               0           Coffee Shop   Japanese Restaurant   
3 -79.360636               0           Coffee Shop                  Café   
4 -79.378937               0           Coffee Shop        Clothing Store   

  3rd Most Common Venue 4th Most Common Venue      5th Most Common Venue  \
0                 Trail              Building              Women's Store   
1                Bakery                  Café                        Pub   
2      Sushi Restaurant               Gay Bar                 Restaurant   
3                  Park                   Pub                     Bakery   
4        Cosmetics Shop                  Café  Middle Eastern Restaurant   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0        Discount Store           Event Space  Ethiopian Restaurant   
1     Convenience Store    Italian Restaurant           Pizza Place   
2                 Hotel           Men's Store                   Pub   
3    Mexican Restaurant               Theater            Restaurant   
4            Restaurant   Japanese Restaurant    Italian Restaurant   

  9th Most Common Venue       10th Most Common Venue  
0     Electronics Store  Eastern European Restaurant  
1            Restaurant                Deli / Bodega  
2       Bubble Tea Shop                          Gym  
3        Breakfast Spot                      Brewery  
4        Ice Cream Shop              Bubble Tea Shop

### Finally, let's visualize the resulting clusters

In [44]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(dt_merged['Latitude'], dt_merged['Longitude'], dt_merged['Neighbourhood'], dt_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

### Examine the Clusters

In [45]:
dt_merged.loc[dt_merged['Cluster Labels'] == 0, dt_merged.columns[[2] + list(range(5, dt_merged.shape[1]))]]

Neighbourhood  Cluster Labels  \
1                         Cabbagetown, St. James Town               0   
2                                Church and Wellesley               0   
3                           Harbourfront, Regent Park               0   
4                            Ryerson, Garden District               0   
5                                      St. James Town               0   
6                                         Berczy Park               0   
7                                  Central Bay Street               0   
8                            Adelaide, King, Richmond               0   
9   Harbourfront East, Toronto Islands, Union Station               0   
10           Design Exchange, Toronto Dominion Centre               0   
11                     Commerce Court, Victoria Hotel               0   
15                    Stn A PO Boxes 25 The Esplanade               0   
16             First Canadian Place, Underground city               0   

   1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
1            Coffee Shop                  Park                Bakery   
2            Coffee Shop   Japanese Restaurant      Sushi Restaurant   
3            Coffee Shop                  Café                  Park   
4            Coffee Shop        Clothing Store        Cosmetics Shop   
5                  Hotel    Italian Restaurant           Coffee Shop   
6            Coffee Shop          Cocktail Bar                  Café   
7            Coffee Shop    Italian Restaurant        Ice Cream Shop   
8            Coffee Shop                  Café                   Bar   
9            Coffee Shop                 Hotel              Aquarium   
10           Coffee Shop                  Café                 Hotel   
11           Coffee Shop                  Café                 Hotel   
15           Coffee Shop            Restaurant                  Café   
16           Coffee Shop                  Café                 Hotel   

   4th Most Common Venue      5th Most Common Venue 6th Most Common Venue  \
1                   Café                        Pub     Convenience Store   
2                Gay Bar                 Restaurant                 Hotel   
3                    Pub                     Bakery    Mexican Restaurant   
4                   Café  Middle Eastern Restaurant            Restaurant   
5                   Café                 Restaurant          Cocktail Bar   
6             Steakhouse                   Beer Bar    Seafood Restaurant   
7         Sandwich Place               Burger Joint                  Café   
8             Steakhouse            Thai Restaurant                   Gym   
9                   Café         Italian Restaurant   Fried Chicken Joint   
10            Restaurant                  Gastropub   American Restaurant   
11   American Restaurant                 Restaurant             Gastropub   
15                 Hotel         Seafood Restaurant              Beer Bar   
16            Restaurant                 Steakhouse   American Restaurant   

   7th Most Common Venue      8th Most Common Venue 9th Most Common Venue  \
1     Italian Restaurant                Pizza Place            Restaurant   
2            Men's Store                        Pub       Bubble Tea Shop   
3                Theater                 Restaurant        Breakfast Spot   
4    Japanese Restaurant         Italian Restaurant        Ice Cream Shop   
5                 Bakery             Cosmetics Shop        Clothing Store   
6                 Bakery                Cheese Shop        Farmers Market   
7             Restaurant  Middle Eastern Restaurant                   Spa   
8                  Hotel        American Restaurant            Restaurant   
9         Scenic Lookout                     Bakery   Sporting Goods Shop   
10         Deli / Bodega                     Bakery                   Gym   
11    Seafood Restaurant                     Bakery            Steakhouse   
15   

In [46]:
print("The above neighborhood cluster is defined by:" + "coffee shops")

The above neighborhood cluster is defined by:coffee shops


In [47]:
dt_merged.loc[dt_merged['Cluster Labels'] == 1, dt_merged.columns[[2] + list(range(5, dt_merged.shape[1]))]]

Neighbourhood  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0      Rosedale               1                  Park            Playground   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                 Trail              Building         Women's Store   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0        Discount Store           Event Space  Ethiopian Restaurant   

  9th Most Common Venue       10th Most Common Venue  
0     Electronics Store  Eastern European Restaurant

In [48]:
print("The above neighborhood cluster is defined by:" + "Park")

The above neighborhood cluster is defined by:Park


In [49]:
dt_merged.loc[dt_merged['Cluster Labels'] == 2, dt_merged.columns[[2] + list(range(5, dt_merged.shape[1]))]]

Neighbourhood  Cluster Labels  \
12             Harbord, University of Toronto               2   
13  Chinatown, Grange Park, Kensington Market               2   

   1st Most Common Venue          2nd Most Common Venue 3rd Most Common Venue  \
12                  Café                            Bar   Japanese Restaurant   
13                  Café  Vegetarian / Vegan Restaurant    Chinese Restaurant   

   4th Most Common Venue 5th Most Common Venue  6th Most Common Venue  \
12             Bookstore            Restaurant                 Bakery   
13    Mexican Restaurant   Dumpling Restaurant  Vietnamese Restaurant   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
12    Chinese Restaurant              Beer Bar            Beer Store   
13                Bakery                   Bar           Coffee Shop   

   10th Most Common Venue  
12         Sandwich Place  
13            Record Shop

In [50]:
print("The above neighborhood cluster is defined by:" + "Cafe")

The above neighborhood cluster is defined by:Cafe


In [51]:
dt_merged.loc[dt_merged['Cluster Labels'] == 3, dt_merged.columns[[2] + list(range(5, dt_merged.shape[1]))]]

Neighbourhood  Cluster Labels  \
14  CN Tower, Bathurst Quay, Island airport, Harbo...               3   

   1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
14        Airport Lounge      Airport Terminal       Airport Service   

   4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
14         Boat or Ferry      Sculpture Garden                 Plane   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
14           Coffee Shop          Airport Gate       Harbor / Marina   

   10th Most Common Venue  
14                Airport

In [52]:
print("The above neighborhood cluster is defined by:" + "Airport Lounge")

The above neighborhood cluster is defined by:Airport Lounge


In [53]:
dt_merged.loc[dt_merged['Cluster Labels'] == 4, dt_merged.columns[[2] + list(range(5, dt_merged.shape[1]))]]

Neighbourhood  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
17      Christie               4                  Café         Grocery Store   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
17                  Park     Convenience Store             Nightclub   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
17            Baby Store            Restaurant                 Diner   

   9th Most Common Venue 10th Most Common Venue  
17    Italian Restaurant            Coffee Shop

In [54]:
print("The above neighborhood cluster is defined by:" + "grocery stores")

The above neighborhood cluster is defined by:grocery stores
